### PR TITLE
[FLINK-19232][python] Support MapState and MapView in PyFlink.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request supports MapState and MapView in PyFlink (The support of iterating a MapState/MapView will be contained in next PR).*

## Brief change log

  - *Support MapState in PyFlink.*
  - *Support MapView in PyFlink.*

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java [hunk 1]
```diff
@@ -636,5 +670,173 @@ private static StateRequestHandler getStateRequestHandler(
...
+		}
+
+		private CompletionStage<BeamFnApi.StateResponse.Builder> handleMapGetRequest(BeamFnApi.StateRequest request)
+				throws Exception {
+			// The structure of get request bytes is:
+			// [flag(1 byte)][serialized map key(if needed)]
+			byte[] getRequest = request.getGet().getContinuationToken().toByteArray();
+			byte getFlag = getRequest[0];
+			if (reuseByteArrayWrapper == null) {
+				reuseByteArrayWrapper = new ByteArrayWrapper(getRequest, 1);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/streaming/api/runners/python/beam/BeamPythonFunctionRunner.java:711]: what about initializing reuseByteArrayWrapper in the constructor?

File: flink-python/src/main/java/org/apache/flink/python/PythonOptions.java [hunk 2]
```diff
@@ -173,4 +173,24 @@
...
+			+ "Note that this is an experimental flag and might not be available in future releases.");
+
+	/**
+	 * The maximum number of write requests cached in a Python MapState.
+	 */
+	@Experimental
+	public static final ConfigOption<Integer> MAP_STATE_WRITE_CACHE_SIZE = ConfigOptions
+		.key("python.map-state.write.cache.size")
+		.defaultValue(1000)
+		.withDescription("The maximum number of write requests cached in a Python MapState. Note that this " +
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/PythonOptions.java:194]:
```suggestion
		.withDescription("The maximum number of cached write requests for a single Python MapState. The write requests will be flushed to the state backend (managed in the Java operator) when the number of cached write requests exceed this limit. Note that this " +
```

File: flink-python/pyflink/fn_execution/state_impl.py [hunk 3]
```diff
@@ -119,13 +121,331 @@ def clear(self):
...
+        self._is_empty = True
+        self._write_cache.clear()
+
+    def commit(self):
+        to_await = None
+        if self._cleared:
+            to_await = self._map_state_handler.clear(self._state_key)
+        if self._write_cache:
+            append_items = []
+            for map_key, existed_and_value in self._write_cache.items():
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/state_impl.py:381]:
```suggestion
            for map_key, (exists, value) in self._write_cache.items():
```

File: flink-python/pyflink/fn_execution/state_impl.py [hunk 4]
```diff
@@ -119,13 +121,331 @@ def clear(self):
...
+    def get(self, map_key):
+        if map_key in self._write_cache:
+            existed_and_value = self._write_cache[map_key]
+            if existed_and_value[0]:
+                return existed_and_value[1]
+            else:
+                raise KeyError("Mapping key %s not found!" % map_key)
+        if self._cleared:
+            raise KeyError("Mapping key %s not found!" % map_key)
+        existed, value = self._map_state_handler.blocking_get(
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/state_impl.py:332]:
```suggestion
        exists, value = self._map_state_handler.blocking_get(
```

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your comments! I have updated this PR.

**Pull request closed without merging**